src/Microsoft.Azure.WebJobs.Host/JobHost.cs

```diff
@@ -329,38 +335,104 @@ private static IFunctionDefinition ResolveFunctionDefinition(MethodInfo method,
+        // When this task is signalled, _context is initialized. 
+        private Task EnsureHostStartedAsync(CancellationToken cancellationToken)
         {
-            JobHostContext context = await _config.CreateAndLogHostStartedAsync(this, _shutdownTokenSource.Token, cancellationToken);
+            if (_context != null)
+            {
+                return Task.CompletedTask;
+            }
+
+            TaskCompletionSource<bool> tsc = null;
 
-            lock (_contextLock)
+            lock (_stopTaskLock)
             {
-                if (_context == null)
+                if (_initializationRunning == null)
                 {
-                    _context = context;
-                    _listener = context.Listener;
+                    // This thread wins the race and owns initialing. 
+                    tsc = new TaskCompletionSource<bool>();
+                    _initializationRunning = tsc.Task;
                 }
             }
 
-            _logger = _context.LoggerFactory?.CreateLogger(LogCategories.Startup);
+            if (tsc != null)
+            {
+                // Ignore the return value and use tsc so that all threads are awaiting the same thing. 
+                Task ignore = RuntimeInitAsync(cancellationToken, tsc);
+            }
 
-            return _context;
+            return _initializationRunning;
         }
 
-        private Task EnsureHostStartedAsync(CancellationToken cancellationToken)
+        // Caller gaurantees this is single-threaded. 
+        // Set initializationTask when complete, many threads can wait on that. 
+        // When complete, the fields should be initialized to allow runtime usage. 
+        private async Task RuntimeInitAsync(CancellationToken cancellationToken, TaskCompletionSource<bool> initializationTask)
         {
-            return LazyInitializer.EnsureInitialized<Task<JobHostContext>>(ref _contextTask,
-                ref _contextTaskInitialized,
-                ref _contextTaskLock,
-                () => CreateContextAndLogHostStartedAsync(cancellationToken));
+            try
+            {
+                // Do real initialization 
+                PopulateStaticServices();
+
+                JobHostContext context = await _config.CreateJobHostContextAsync(_services, this, _shutdownTokenSource.Token, cancellationToken);
+
+                _context = context;
+                _listener = context.Listener;
+                _logger = _context.LoggerFactory?.CreateLogger(LogCategories.Startup);
+
+                initializationTask.SetResult(true);
+            }
+            catch (Exception e)
+            {
+                initializationTask.SetException(e);
+            }
         }
 
-        private void ThrowIfDisposed()
+        // Ensure the static services are initialized. 
+        // These are derived from the underlying JobHostConfiguration. 
+        private void PopulateStaticServices()
```

Comment:
It looks like this can be called from several places -- do we need to lock around this to make sure it's thread-safe?